File: Overlord.Core/Overlord.Core/Models/CombatModels.cs
using System.Text.Json.Serialization;

namespace Overlord.Core.Models;

/// <summary>
/// Represents a ground battle between platoons.
/// </summary>
public class Battle
{
    public int PlanetID { get; set; }
    public string PlanetName { get; set; } = string.Empty;
    public FactionType AttackerFaction { get; set; }
    public FactionType DefenderFaction { get; set; }
    public List<int> AttackingPlatoonIDs { get; set; } = new List<int>();
    public List<int> DefendingPlatoonIDs { get; set; } = new List<int>();
}

/// <summary>
/// Result of a ground battle.
/// </summary>
public class BattleResult
{
    public bool AttackerWins { get; set; }
    public int AttackerStrength { get; set; }
    public int DefenderStrength { get; set; }
    public int AttackerCasualties { get; set; }
    public int DefenderCasualties { get; set; }
    public bool PlanetCaptured { get; set; }
    public Dictionary<int, int> PlatoonCasualties { get; set; } = new Dictionary<int, int>();
}

/// <summary>
/// Represents a space battle between craft fleets.
/// </summary>
public class SpaceBattle
{
    public int PlanetID { get; set; }
    public string PlanetName { get; set; } = string.Empty;
    public FactionType AttackerFaction { get; set; }
    public FactionType DefenderFaction { get; set; }
    public List<int> AttackerCraftIDs { get; set; } = new List<int>();
    public List<int> DefenderCraftIDs { get; set; } = new List<int>();
    public bool DefenderHasOrbitalDefense { get; set; }
}

/// <summary>
/// Result of a space battle.
/// </summary>
public class SpaceBattleResult
{
    public bool AttackerWins { get; set; }
    public int AttackerStrength { get; set; }
    public int DefenderStrength { get; set; }
    public int DamagePerCraft { get; set; }
    public List<int> DestroyedCraftIDs { get; set; } = new List<int>();
    public Dictionary<int, int> CraftDamage { get; set; } = new Dictionary<int, int>();
}

/// <summary>
/// Result of planetary bombardment.
/// </summary>
public class BombardmentResult
{
    public int PlanetID { get; set; }
    public int BombardmentStrength { get; set; }
    public int StructuresDestroyed { get; set; }
    public List<int> DestroyedStructureIDs { get; set; } = new List<int>();
    public int CivilianCasualties { get; set; }
    public int NewMorale { get; set; }
}

/// <summary>
/// Result of planetary invasion.
/// </summary>
public class InvasionResult
{
    public bool AttackerWins { get; set; }
    public bool PlanetCaptured { get; set; }
    public bool InstantSurrender { get; set; }
    public int AttackerCasualties { get; set; }
    public int DefenderCasualties { get; set; }
    public int NewPopulation { get; set; }
    public int NewMorale { get; set; }
    public ResourceDelta? CapturedResources { get; set; }
}

File: Overlord.Phaser/src/core/models/CombatModels.ts
import { FactionType } from './Enums';
import { ResourceDelta } from './ResourceModels';

/**
 * Represents a ground battle between platoons.
 */
export class Battle {
  public planetID: number = 0;
  public planetName: string = '';
  public attackerFaction: FactionType = FactionType.Player;
  public defenderFaction: FactionType = FactionType.Neutral;
  public attackingPlatoonIDs: number[] = [];
  public defendingPlatoonIDs: number[] = [];
}

/**
 * Result of a ground battle.
 */
export class BattleResult {
  public attackerWins: boolean = false;
  public attackerStrength: number = 0;
  public defenderStrength: number = 0;
  public attackerCasualties: number = 0;
  public defenderCasualties: number = 0;
  public planetCaptured: boolean = false;
  public platoonCasualties: Map<number, number> = new Map();
}

/**
 * Represents a space battle between craft fleets.
 */
export class SpaceBattle {
  public planetID: number = 0;
  public planetName: string = '';
  public attackerFaction: FactionType = FactionType.Player;
  public defenderFaction: FactionType = FactionType.Neutral;
  public attackerCraftIDs: number[] = [];
  public defenderCraftIDs: number[] = [];
  public defenderHasOrbitalDefense: boolean = false;
}

/**
 * Result of a space battle.
 */
export class SpaceBattleResult {
  public attackerWins: boolean = false;
  public attackerStrength: number = 0;
  public defenderStrength: number = 0;
  public damagePerCraft: number = 0;
  public destroyedCraftIDs: number[] = [];
  public craftDamage: Map<number, number> = new Map();
}

/**
 * Result of planetary bombardment.
 */
export class BombardmentResult {
  public planetID: number = 0;
  public bombardmentStrength: number = 0;
  public structuresDestroyed: number = 0;
  public destroyedStructureIDs: number[] = [];
  public civilianCasualties: number = 0;
  public newMorale: number = 0;
}

/**
 * Result of planetary invasion.
 */
export class InvasionResult {
  public attackerWins: boolean = false;
  public planetCaptured: boolean = false;
  public instantSurrender: boolean = false;
  public attackerCasualties: number = 0;
  public defenderCasualties: number = 0;
  public newPopulation: number = 0;
  public newMorale: number = 0;
  public capturedResources: ResourceDelta | null = null;
}
